File: apps/web/lib/api/retrieval.ts
import { apiMutation, hasString, isArrayOf, isRecord } from "./client";
import type {
  DocumentIndexSummary,
  EvidenceContext,
  RetrievalSearchResponse,
} from "@/lib/types/api";

function isDocumentIndexSummary(value: unknown): value is DocumentIndexSummary {
  return isRecord(value) && hasString(value, "document_id") && value.status === "INDEXED" && typeof value.chunk_count === "number";
}

function isEvidenceContext(value: unknown): value is EvidenceContext {
  return isRecord(value) && typeof value.score === "number" && value.similarity_label === "retrieval similarity" && hasString(value, "document_id") && hasString(value, "filename") && typeof value.page_number === "number" && typeof value.text === "string" && typeof value.chunk_index === "number";
}

function isRetrievalSearchResponse(value: unknown): value is RetrievalSearchResponse {
  return isRecord(value) && hasString(value, "query") && isArrayOf(value.results, isEvidenceContext) && typeof value.indexed_document_count === "number";
}

export function indexDocument(documentId: string): Promise<DocumentIndexSummary> {
  return apiMutation(`/api/v1/documents/${encodeURIComponent(documentId)}/index`, "POST", isDocumentIndexSummary);
}

export function searchEvidence(profileId: string, query: string, topK = 5): Promise<RetrievalSearchResponse> {
  return apiMutation(`/api/v1/profiles/${encodeURIComponent(profileId)}/retrieval/search`, "POST", isRetrievalSearchResponse, { query, top_k: topK });
}

File: apps/api/app/schemas/retrieval.py
"""Indexing and semantic-evidence API contracts."""

from typing import Literal

from pydantic import Field, field_validator

from app.database.enums import VectorIndexStatus
from app.schemas.common import SchemaModel


class DocumentIndexSummary(SchemaModel):
    document_id: str
    status: VectorIndexStatus
    chunk_count: int


class RetrievalSearchRequest(SchemaModel):
    query: str = Field(min_length=1, max_length=500)
    top_k: int = Field(default=5, ge=1, le=20)

    @field_validator("query")
    @classmethod
    def query_must_contain_text(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("Query must contain text.")
        return value


class EvidenceContext(SchemaModel):
    score: float
    similarity_label: Literal["retrieval similarity"] = "retrieval similarity"
    document_id: str
    filename: str
    page_number: int
    text: str
    chunk_index: int


class RetrievalSearchResponse(SchemaModel):
    query: str
    results: list[EvidenceContext]
    indexed_document_count: int
